### PR TITLE
Universal shadowserver parser

### PR DESCRIPTION
**Pull request by dmth**
This PR introduces a new strategy to parse Shadowserver feeds.

It's a tidied version of https://github.com/certtools/intelmq/pull/548

**Known ToDos:**
 * Add Conversion Functions for some Data, as @sebix already suggested
  * IPMI: `*_auth`, `usernames`, `nulluser` and `anon_login` replace  yes/no with true/false
  * IPMI: `default_kg` replace the - or default  with None
  * Hostnames can contain IP's which has to be handled.
 * Add this Bot to the documentation or update the documentation.

### CODE REVIEW

**Comment by sebix** [comment]
Current behavior for extra:

    'extra': '{"sic": 654321, "response_size": 116, "naics": 123456}',
new one:

    'extra': '{"shadowserver": {"naics": "123456", "sic": "654321", "size": ' '"116", "tag": "chargen"}}',

If we switch to this new behavior, we also should change the documentation about this field.

**Comment by dmth** [comment]
> If we switch to this new behavior, we also should change the documentation about this field.

Correct. Which documentation do you refer to?

File: intelmq/bots/parsers/shadowserver/parser.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+            # fields that coul not be added to the standard intelmq fields
+            # the parser is going to write this information into an object
+            # one level below the "extra root"
+            # e.g.: extra {'shadowserver': {'cc_dns': '127.0.0.1'}
+
+            # set classification.type
+            # TODO this might get dynamic in some feeds.
+            # How to handle that?
+
+            if not conf.get('classification_type'):
```
Comment (sebix) [inline intelmq/bots/parsers/shadowserver/parser.py]: Why is this not in init?

File: intelmq/bots/parsers/shadowserver/parser.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        raw_report = utils.base64_decode(report["raw"])
+        csvr = csv.DictReader(io.StringIO(raw_report))
+
+        # create an array of fieldnames,
+        # those were automagically created by the dictreader
+        allfields = csvr.fieldnames
+
+        # Set a switch if the parser shall reset the feed.name,
+        # code and feedurl for this event
+        self.override = False
```
Comment (sebix) [inline intelmq/bots/parsers/shadowserver/parser.py]: Why is this not in init?

Comment (dmth) [inline intelmq/bots/parsers/shadowserver/parser.py]: Putting this to init should not have side effects.

File: intelmq/bots/parsers/shadowserver/parser.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+            if not conf.get('classification_type'):
+                self.logger.warn("The classification type for "
+                    + self.parameters.feedname
+                    + " could not be determined. Check ParserConfig!")
+            else:
+                event.add('classification.type', conf.get('classification_type'))
+
+            # set feed.name and url, honor the override parameter
+
+            if 'feed.url' in event and self.override:
```
Comment (dmth) [inline intelmq/bots/parsers/shadowserver/parser.py]: I doubt that the if-clause is correct in this place. It should be:
If "field is present and override is set **else if** field is not present"

this should also be corrected for the next two parameters.

Comment (sebix) [inline intelmq/bots/parsers/shadowserver/parser.py]: Already catched that one, otherwise a KeyExists would be thrown.

**Comment by aaronkaplan** [comment]
Can we close this PR? I believe the discussion moved to PR #555 ...

**Comment by sebix** [comment]
I'm not keen to close it before the discussion ended. #549 will be closed automatically if #555 is merged.

**Comment by aaronkaplan** [comment]
All right. We'll keep it open then. 

---
Mobile

**Pull request closed without merging**